wallet-new/integration/Types.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE ConstraintKinds #-}
+
+module Types
+    ( Probability
+    , createProbability
+    , getProbability
+    , Action (..)
+    , WalletState (..)
+    , TestResult (..)
+    , ActionWalletState
+    , ActionProbabilities
+    , WIntTestMode
+    ) where
+
+import Universum
+
+import Cardano.Wallet.API.V1.Types (Wallet)
+
+type WIntTestMode m =
+    ( MonadIO m
+    , MonadThrow m
+    )
+
+-- | The probability type that captures the chance of
+-- our random action selection choosing a specific @Action@.
+-- The number is restricted to 1 - 100 - see @createProbability@.
+data Probability = Probability { getProbability :: Int }
+    deriving (Show, Eq)
+
+
+-- | Safe constructor that checks the values.
+createProbability :: Int -> Probability
+createProbability prob
+    -- TODO(ks): Maybe we should @throwM@ or @either@ here.
+    | prob <= 0 || prob > 100 =
+        error "Probability should be between 1 - 100."
+    | otherwise               = Probability prob
+
+
+-- | Actions that can be called from the test.
+data Action
+    = CreateWallet
+    | GetWallet
+    | CreateAccount
+    | CreateTransaction
+    | GetTransaction
+    deriving (Show, Eq)
+
+
+-- | The type that defines the probabilites of the actions
+-- TODO(ks): We could create a custom constructor with valid
+-- values only.
+-- Add invariant?
+type ActionProbabilities = [(Action, Probability)]
```

Comment:
This is a partial lookup-table -- why not `type ActionProbabilities = Action -> Probability`? Constructing them is nearly as simple, with a guarantee of totality. 

```haskell
foo :: ActionProbabilities 
foo = Probability . \case
  CreateWallet -> 3
  -- etc..
```

Reply:
Interesting, but AFAIK we don't require a guarantee of totality, we just need to check if the probabilities add up to 100 (comment above).

Reply:
Another thing to note is `QuickCheck`'s `frequency` function -- it uses something like:

```
frequency :: [(Int, Gen a)] -> Gen a
frequency = join . elements . concatMap (\(i, g) -> replicate i g)
```

which essentially uses the sum of the weights in the list to normalize. This eliminates partiality *and* the requirement that the probability sum to 100.